### PR TITLE
Visualize Workflows

### PR DESCRIPTION
**Pull request by Devessier**
## Features

- Fetch a workflow and display it in a tree with the React Flow library
- The nodes are positioned by an algorithm
- The feature is put behind a feature flag. The `/workflow/:id` route is disabled if the flag is off.
- I started implementing a right drawer. That's a big WIP and it will be finished in another PR.

## How to test this feature

1. Create a workflow instance in the database through a GraphQL query. See below for instructions.
2. After enabling the feature flag, you should be able to see the workflow you created in the workflows list. To visualize the workflow, go to the `/workflow/:id` page where the id is the id of the workflow. See the video for a quick way to do so.

```gql
// First
mutation createWorkflow($data: WorkflowCreateInput!) {
  createWorkflow(data: $data) {
    id
  }
}

// Result
{
  "data": {
    "name": "test"
  }
}

// Second
mutation createWorkflowVersion($data: WorkflowVersionCreateInput!) {
  createWorkflowVersion (data: $data) {
    id
  }
}

// Result
{
  "data": {
    "name": "v1",
    "trigger": {
  "name": "trigger",
"displayName": "New or Updated Row",
"type": "DATABASE_EVENT",
"settings": {
"eventName": "company.created",
"triggerName": "Company Created"
},
  "nextAction": {
    "name": "step_1",
    "displayName": "Code",
    "type": "CODE",
    "valid": true,
    "settings": {
      "serverlessFunctionId": "function_id",
      "errorHandlingOptions": {
        "retryOnFailure": {
          "value": false
        },
        "continueOnFailure": {
          "value": false
        }
      }
    }
  }
},
"workflowId": "workflow_id"
  }
}
```

https://github.com/user-attachments/assets/42bbd98c-5e13-447c-9307-461a18ac2195

### CODE REVIEW

File: packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+const EMPTY_FLOW_DATA: WorkflowDiagram = {
+  nodes: [],
+  edges: [],
+};
+
+const generateId = () => {
+  return Math.random().toString(16).slice(2);
+};
+
+const workflowTriggerToFlow = (trigger: WorkflowTrigger): WorkflowDiagram => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx:30]: naming: generateWorklowDiagram

File: packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  const nodes: Array<Node<WorkflowDiagramNodeData>> = [];
+  const edges: Array<Edge> = [];
+
+  // Helper function to generate nodes and edges recursively
+  const generateFlow = (
```
Comment (charlesBochet) [inline packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx:35]: processNode

File: packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+type WorkflowShowPageEffectProps = {
+  workflowId: string;
+};
+
+const EMPTY_FLOW_DATA: WorkflowDiagram = {
+  nodes: [],
+  edges: [],
+};
+
+const generateId = () => {
```
Comment (charlesBochet) [inline packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx:26]: use v4()

File: packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+      id: generateId(),
+      source: parentNodeId,
+      target: nodeId,
+      markerEnd: {
+        type: MarkerType.ArrowClosed,
+      },
+    });
+
+    // Recursively generate flow for the next action if it exists
+    if (action.nextAction !== undefined) {
```
Comment (charlesBochet) [inline packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx:65]: use utils: 
isDefined(action.nextAction)
(everywhere)

File: packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+  const lastVersion = workflow.versions[0];
+  if (lastVersion === undefined || lastVersion.trigger === undefined) {
+    return EMPTY_FLOW_DATA;
+  }
+
+  return workflowTriggerToFlow(lastVersion.trigger);
+};
+
+const addCreateStepNodes = (
```
Comment (charlesBochet) [inline packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx:110]: utils

File: packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  nodes: Array<Node<WorkflowDiagramNodeData>>,
+  edges: Array<Edge>,
+) => {
+  const nodesWithoutTargets = nodes.filter((node) =>
+    edges.every((edge) => edge.source !== node.id),
+  );
+
+  const updatedNodes: typeof nodes = nodes.slice();
```
Comment (charlesBochet) [inline packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx:118]: use types directly

File: packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    objectRecordId: workflowId,
+    recordGqlFields: {
+      id: true,
+      name: true,
+      versions: true,
+      publishedVersionId: true,
+    },
+  });
+
+  const flowLastVersion = useMemo(
```
Comment (charlesBochet) [inline packages/twenty-front/src/pages/workflows/WorkflowShowPageEffect.tsx:165]: useMemo does not seem useful here

File: packages/twenty-front/src/pages/workflows/nodes/StepNode.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  .selectable.selected &,
+  .selectable:focus &,
+  .selectable:focus-visible & {
+    background-color: ${({ theme }) => theme.color.blue};
+    color: ${({ theme }) => theme.font.color.inverted};
+  }
+`;
+
+const StyledStepNodeInnerContainer = styled.div`
```
Comment (charlesBochet) [inline packages/twenty-front/src/pages/workflows/nodes/StepNode.tsx:36]: extract in multiple files

**Review by charlesBochet — APPROVED** [review]

File: packages/twenty-front/src/modules/ui/layout/right-drawer/constants/RightDrawerPageTitles.ts [hunk 9]
```diff
@@ -5,4 +5,5 @@ export const RIGHT_DRAWER_PAGE_TITLES = {
   [RightDrawerPages.ViewCalendarEvent]: 'Calendar Event',
   [RightDrawerPages.ViewRecord]: 'Record Editor',
   [RightDrawerPages.Copilot]: 'Copilot',
+  [RightDrawerPages.Workflow]: 'Workflow',
```
[the remark this answers is not in the document]
Comment (Devessier) [reply packages/twenty-front/src/modules/ui/layout/right-drawer/constants/RightDrawerPageTitles.ts:8]: I will fix this issue in my next PR as this PR doesn't implement much things about the Right Drawer.

Things that remain to do:

- Don't hesitate to create many drawers. Create a drawer per view.

File: packages/twenty-front/src/modules/workflow/states/currentWorkflowDataState.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import { WorkflowDiagram } from '@/workflow/types/Workflow';
+import { createState } from 'twenty-ui';
+
+export const currentWorkflowDataState = createState<
```
[the remark this answers is not in the document]
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/states/currentWorkflowDataState.ts:4]: I will rename the state to `showPageWorkflowDiagramState ` for now. Later, we'll make it a component state.

File: packages/twenty-front/src/pages/workflows/WorkflowShowPage.tsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+const LoadedWorkflow = ({ flowData }: { flowData: WorkflowDiagram }) => {
+  const [rawNodes, setRawNodes, onRawNodesChange] = useNodesState(
+    flowData.nodes,
+  );
+  const [rawEdges, setRawEdges, onRawEdgesChange] = useEdgesState(
+    flowData.edges,
+  );
+
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (Devessier) [reply packages/twenty-front/src/pages/workflows/WorkflowShowPage.tsx:86]: I found a way to remove it.

File: packages/twenty-front/src/modules/workflow/components/WorkflowShowPageEffect.tsx [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+type WorkflowShowPageEffectProps = {
+  workflowId: string;
+};
+
+const EMPTY_FLOW_DATA: WorkflowDiagram = {
+  nodes: [],
+  edges: [],
+};
+
+const getFlowLastVersion = (
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/workflow/components/WorkflowShowPageEffect.tsx:23]: extract this to utils (and test it!)

File: packages/twenty-front/src/modules/workflow/components/nodes/CreateStepNode.tsx [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import { IconButton } from '@/ui/input/button/components/IconButton';
```
Comment (charlesBochet) [inline packages/twenty-front/src/modules/workflow/components/nodes/CreateStepNode.tsx:1]: by project convention when we create subfolder we do it the other way: packages/twenty-front/src/modules/workflow/nodes/components/CreateStepNode.tsx

The idea is that all leaves folder respect the same structure of components/hooks/types, ...

**Review by charlesBochet — APPROVED** [review]
Bravo @Devessier, this is a good first one! and huge thanks for taking the time to write code in our way!

**Comment by charlesBochet** [comment]
@Devessier the CI is red: https://github.com/twentyhq/twenty/actions/runs/10512071157/job/29124522652?pr=6697

Locally: 
- npx nx run twenty-front:typecheck

Then, (can be done in another PR), you'll have to add storybook tests:
- npx nx run twenty-front:storybook:test --configuration=pages (high level, we only render Pages for pixel diff and make sure they run)
- npx nx run twenty-front:storybook:test --configuration=modules (ideally we want one story per component of group of component, it's integration testing)

And unit tests on your utils / hooks:
- npx nx run twenty-front:test

File: packages/twenty-front/src/modules/workflow/components/nodes/CreateStepNode.tsx (hunk 13 above)
Comment (Devessier) [reply packages/twenty-front/src/modules/workflow/components/nodes/CreateStepNode.tsx:1]: I chose to flatten the components directory instead. I don't think having a `nodes` submodule is essential.

**Pull request merged**